File: src/hooks/useMicrophone.ts
import { useState, useCallback } from 'react';

export const useMicrophone = () => {
  const [isPermissionGranted, setIsPermissionGranted] = useState(false);
  const [isPermissionDenied, setIsPermissionDenied] = useState(false);
  const [isRecording, setIsRecording] = useState(false);
  const [isPlaying, setIsPlaying] = useState(false);
  const [audioChunks, setAudioChunks] = useState<Blob[]>([]);
  const [audioContext, setAudioContext] = useState<AudioContext | null>(null);
  const [mediaRecorder, setMediaRecorder] = useState<MediaRecorder | null>(null);

  const requestMicrophonePermission = useCallback(async () => {
    try {
      const stream = await navigator.mediaDevices.getUserMedia({ audio: true });
      stream.getTracks().forEach(track => track.stop());
      setIsPermissionGranted(true);
      setIsPermissionDenied(false);
      return true;
    } catch (error) {
      console.error('Error requesting microphone permission:', error);
      setIsPermissionDenied(true);
      return false;
    }
  }, []);

  const startRecording = useCallback(async () => {
    if (!isPermissionGranted) {
      const granted = await requestMicrophonePermission();
      if (!granted) return;
    }

    try {
      const stream = await navigator.mediaDevices.getUserMedia({ audio: true });
      const recorder = new MediaRecorder(stream);
      const chunks: Blob[] = [];

      recorder.ondataavailable = (e) => {
        if (e.data.size > 0) {
          chunks.push(e.data);
        }
      };

      recorder.onstop = () => {
        setAudioChunks(chunks);
        stream.getTracks().forEach(track => track.stop());
      };

      recorder.start();
      setMediaRecorder(recorder);
      setIsRecording(true);
    } catch (error) {
      console.error('Error starting recording:', error);
    }
  }, [isPermissionGranted, requestMicrophonePermission]);

  const stopRecording = useCallback(() => {
    if (mediaRecorder && isRecording) {
      mediaRecorder.stop();
      setIsRecording(false);
    }
  }, [mediaRecorder, isRecording]);

  const playAudio = useCallback(async () => {
    if (audioChunks.length === 0) return;

    try {
      const audioBlob = new Blob(audioChunks, { type: 'audio/webm' });
      const audioUrl = URL.createObjectURL(audioBlob);
      const audio = new Audio(audioUrl);
      
      setIsPlaying(true);
      await audio.play();
      
      audio.onended = () => {
        setIsPlaying(false);
        URL.revokeObjectURL(audioUrl);
      };
    } catch (error) {
      console.error('Error playing audio:', error);
      setIsPlaying(false);
    }
  }, [audioChunks]);

  return {
    isPermissionGranted,
    isPermissionDenied,
    isRecording,
    isPlaying,
    requestMicrophonePermission,
    startRecording,
    stopRecording,
    playAudio
  };
}; 

File: lib/hooks/use_microphone.dart
import 'package:flutter_hooks/flutter_hooks.dart';
import 'package:permission_handler/permission_handler.dart';
import 'package:record/record.dart';

class MicrophoneHook {
  final bool isPermissionGranted;
  final bool isPermissionDenied;
  final bool isRecording;
  final Future<bool> Function() requestPermission;
  final Future<void> Function() startRecording;
  final Future<void> Function() stopRecording;

  MicrophoneHook({
    required this.isPermissionGranted,
    required this.isPermissionDenied,
    required this.isRecording,
    required this.requestPermission,
    required this.startRecording,
    required this.stopRecording,
  });
}

MicrophoneHook useMicrophone() {
  final isPermissionGranted = useState(false);
  final isPermissionDenied = useState(false);
  final isRecording = useState(false);
  final recorder = useState<AudioRecorder?>(null);

  Future<bool> requestPermission() async {
    final status = await Permission.microphone.request();
    
    if (status.isGranted) {
      isPermissionGranted.value = true;
      isPermissionDenied.value = false;
      return true;
    } else {
      isPermissionDenied.value = true;
      return false;
    }
  }

  Future<void> startRecording() async {
    if (!isPermissionGranted.value) {
      final granted = await requestPermission();
      if (!granted) return;
    }

    try {
      final record = AudioRecorder();
      await record.start(
        const RecordConfig(),
        path: '${DateTime.now().millisecondsSinceEpoch}.m4a',
      );
      recorder.value = record;
      isRecording.value = true;
    } catch (e) {
      print('Error starting recording: $e');
    }
  }

  Future<void> stopRecording() async {
    if (recorder.value != null && isRecording.value) {
      await recorder.value!.stop();
      isRecording.value = false;
    }
  }

  return MicrophoneHook(
    isPermissionGranted: isPermissionGranted.value,
    isPermissionDenied: isPermissionDenied.value,
    isRecording: isRecording.value,
    requestPermission: requestPermission,
    startRecording: startRecording,
    stopRecording: stopRecording,
  );
} 